x509/src/lib.rs

```diff
@@ -14,12 +14,110 @@ extern crate alloc;
+    V2 = 1,
+
+    /// Version 3
+    V3 = 2,
+}
+
+impl Default for Version {
+    fn default() -> Self {
+        Self::V1
+    }
+}
+
+/// X.509 `TbsCertificate` as defined in [RFC 5280 Section 4.1]
+///
+/// ASN.1 structure containing the names of the subject and issuer, a public
+/// key associated with the subject, a validity period, and other associated
+/// information.
+///
+/// ```text
+/// TBSCertificate  ::=  SEQUENCE  {
+///     version         [0]  EXPLICIT Version DEFAULT v1,
+///     serialNumber         CertificateSerialNumber,
+///     signature            AlgorithmIdentifier,
+///     issuer               Name,
+///     validity             Validity,
+///     subject              Name,
+///     subjectPublicKeyInfo SubjectPublicKeyInfo,
+///     issuerUniqueID  [1]  IMPLICIT UniqueIdentifier OPTIONAL,
+///                          -- If present, version MUST be v2 or v3
+///     subjectUniqueID [2]  IMPLICIT UniqueIdentifier OPTIONAL,
+///                          -- If present, version MUST be v2 or v3
+///     extensions      [3]  Extensions OPTIONAL
+///                          -- If present, version MUST be v3 --
+/// }
+/// ```
+///
+/// [RFC 5280 Section 4.1]: https://datatracker.ietf.org/doc/html/rfc5280#section-4.1
+#[derive(Clone, Debug, Eq, PartialEq, Sequence)]
+#[allow(missing_docs)]
+pub struct TbsCertificate<'a> {
+    /// The certificate version
+    ///
+    /// Note that this value defaults to Version 1 per the RFC. However,
+    /// fields such as `issuer_unique_id`, `subject_unique_id` and `extensions`
+    /// require later versions. Care should be taken in order to ensure
+    /// standards compliance.
+    #[asn1(context_specific = "0", default = "Default::default")]
+    pub version: Version,
+
+    pub serial_number: UIntBytes<'a>,
+    pub signature: AlgorithmIdentifier<'a>,
+    pub issuer: Name<'a>,
+    pub validity: Validity,
+    pub subject: Name<'a>,
+    pub subject_public_key_info: SubjectPublicKeyInfo<'a>,
+
+    #[asn1(context_specific = "1", tag_mode = "IMPLICIT", optional = "true")]
+    pub issuer_unique_id: Option<BitString<'a>>,
+
+    #[asn1(context_specific = "2", tag_mode = "IMPLICIT", optional = "true")]
+    pub subject_unique_id: Option<BitString<'a>>,
+
+    #[asn1(context_specific = "3", tag_mode = "EXPLICIT", optional = "true")]
+    pub extensions: Option<ext::Extensions<'a>>,
+}
+
+/// X.509 certificates are defined in [RFC 5280 Section 4.1].
+///
+/// ```text
+/// Certificate  ::=  SEQUENCE  {
+///     tbsCertificate       TBSCertificate,
+///     signatureAlgorithm   AlgorithmIdentifier,
+///     signature            BIT STRING
+/// }
+/// ```
+///
+/// [RFC 5280 Section 4.1]: https://datatracker.ietf.org/doc/html/rfc5280#section-4.1
+#[derive(Clone, Debug, Eq, PartialEq, Sequence)]
+#[allow(missing_docs)]
+pub struct Certificate<'a> {
+    pub tbs_certificate: TbsCertificate<'a>,
+    pub signature_algorithm: AlgorithmIdentifier<'a>,
+    pub signature: BitString<'a>,
+}
```

Comment:
Hmm, could go either way on moving this stuff into the toplevel module. I think it might be a bit surprising, considering everything else is factored into respectively-named modules.

Reply:
The reason I did that was because these are the primary data types of the crate. Extensions are all under the `ext` module. The same is true with trust anchor types which are under `anchor`. But with x509, the certificate types are the main types of the crate. So I thought it was most ergonomic to put them in the root. I would do the same thing with types that are "groups of certificates" such as `PkiPath`.

Reply:
Also, please note that certificate types are already re-exported in the root of the crate.

Reply:
I think it's somewhat nonidiomatic to have a mixture of types scoped to their own submodules and types defined in `lib.rs`. It's also inconsistent with the other crates in this module, where types are scoped to submodules.

> Extensions are all under the ext module. The same is true with trust anchor types which are under anchor

Yes, but there you are scoping the types to submodules underneath `ext` and `anchor`.

In general I think there are some nice code quality benefits to placing types in respective modules, for example unit tests can be placed in a `mod tests` within each submodule as opposed to a toplevel module which is a big bag of types.

Reply:
I have no interesting in dying on this hill. :)

Are you suggesting to put cert types in a `cert` module? Do you think they should be reexported in the root directory?

Reply:
I'd just keep `Certificate` in the `certificate.rs` it was previously defined in, and put `Version` in a `version.rs`

Reply:
What about `TbsCertificate` and new types like `PkiPath`?

Generally, my strategy has been to keep trivial but related types together. `Version`, `Certificate` and `PkiPath` are all trivial types related to the `TbsCertificate` type. Having a separate file for each seems a bit like overkill. But perhaps not.

`certificate` is rather long. And shortened module names is definitely idiomatic Rust. That's why I proposed `cert`.

Also please note that extensions have a specific re-export strategy. Namely, there are Extension types, which we consider as first-class types and all get defined or re-exported under `ext::pkix`. Then there are related types which are used by the extensions. These are generally not re-exported since they are typically used by simply dereferencing fields on the extensions.

Reply:
I tried reorganizing this and couldn't come up with anything I liked. For example, `use x509::Certificate` seems very ergonomic while `use x509::certificate::Certificate` or even `use x509::cert::Certificate` seems very redundant. Further, trying to split up these related types results in an awkward import structure.

I propose the structure that I just pushed.

1. `lib.rs` is kept clean.
2. redundancy is avoided.
3. closely-related types are grouped by file (this follows the style in the rest of the crate).

Reply:
Many of our other crates re-export types from the toplevel, but define them in modules, keeping those modules private. In fact, that's exactly how it's structured prior to this PR.

I guess I'm just asking for you to put it back the way it was.

Reply:
Aah, looks better now. Maybe keep the module name as `certificate`? It's not part of the public API, so I don't see a reason to abbreviate it.

Reply:
Yeah. Fixing now...

Reply:
Done.